base/util/PKICertImport.bash

```diff
@@ -258,38 +495,316 @@ function PKICertImport() {
+
+        # In the event we only want the leaf certificate, we have to do a
+        # little bit more work...
+        if [ "$PKCS12_LEAF" == "true" ]; then
+            local priv_key="$TMPBASE/cert.key"
+
+            # First, split off only the private key.
+            key_args=("pkcs12")
+            key_args+=("-in" "$CERT_PATH")
+            key_args+=("-clcerts" "-nocerts" "-nodes")
+            key_args+=("-out" "$priv_key")
+
+            if [ "x$arg" != "x$PKCS12_PASSWORD" ]; then
+                # When specified, a path containing the PKCS12 password.
+                key_args+=("-passin" "file:$PKCS12_PASSWORD")
+            fi
+
+            __v openssl "${key_args[@]}"
+            openssl "${key_args[@]}"
+            ret=$?
+            if (( ret != 0 )); then
+                __e "openssl pkcs12 split private key returned: $ret"
+                __secure_rmtmp
+                exit $ret
+            fi
+
+            # Now, join leaf certificate and private key in a new .p12
+            # with the original password.
+            join_args=("pkcs12" "-export")
+            join_args+=("-in" "$PKCS12_CERT_PATH")
+            join_args+=("-inkey" "$priv_key")
+            join_args+=("-out" "$TMPBASE/joined.p12")
+            if [ "x$arg" != "x$PKCS12_PASSWORD" ]; then
+                # When specified, a path containing the PKCS12 password.
+                join_args+=("-passout" "file:$PKCS12_PASSWORD")
+            fi
+
+            __v openssl "${join_args[@]}"
+            openssl "${join_args[@]}"
+            ret=$?
+            if (( ret != 0 )); then
+                __e "openssl pkcs12 join leaf parts returned: $ret"
+                __secure_rmtmp
+                exit $ret
             fi
         fi
     }
 
+    function _import_pkcs12() {
+        local before_certs="$TMPBASE/before_certs.txt"
+        local after_certs="$TMPBASE/after_certs.txt"
+        local before_keys="$TMPBASE/before_keys.txt"
+        local after_keys="$TMPBASE/after_keys.txt"
+
+        # List of known trust flags; combined from man pages and online
+        # documentation.
+        local tf="pPcCTuw"
+
+        # Arguments for listing certificates.
+        local list_args=()
+        list_args+=("-d" "$NSSDB_TYPE$NSSDB")
+
+        # Arguments for import
+        local import_args=("-i" "$CERT_PATH")
+        import_args+=("-d" "$NSSDB_TYPE$NSSDB")
+
+        # When present, specify NSS DB password for listing and importing.
+        if [ "x$NSSDB_PASSWORD" != "x" ]; then
+            list_args+=("-f" "$NSSDB_PASSWORD")
+            import_args+=("-k" "$NSSDB_PASSWORD")
+        fi
+
+        # When present, specify .p12 file password on import.
+        if [ "x$PKCS12_PASSWORD" != "x" ]; then
+            import_args+=("-w" "$PKCS12_PASSWORD")
+        fi
+
+        # List certificates prior to import.
+        __v certutil -L "${list_args[@]}"
+        __v certutil -K "${list_args[@]}"
+        certutil -L "${list_args[@]}" |
```

Comment:
I'm not completely convinced this will work all the time. Perhaps I just need to extend my testbed verification suites, but I'm worried about when the NSS DB / HSM contains a large number of certificates. I think I've seen the HSM truncate the output of either `certutil -L` or `certutil -K`... I guess this is only used on NSS DBs, so that's where we need to worry about it.

Thoughts?

Reply:
Unrelated, but since we're not writing this in C, we don't have an easy way to list just the nickname of new certificates, so this involves parsing the output of the `certutil -L` and `certutil -K` calls. 

To parse the `certutil -L` call,  we look for lines with trust flags (`grep "[$tf]*,[$tf]*,[$tf]*"`), then strip out the trust flags (`sed "s/[[:space:]]*[$tf]*,[$tf]*,[$tf]*//g"`). What remains is the certificate nickname, which may contain spaces. 

To parse the `certutil -K` call, we look for lines beginning with `<$num>` (`grep '^<[[:space:]]*[0-9]*>'`), then strip out that and the key type (`sed 's/^<[[:space:]]*[0-9]*>[[:space:]]*rsa//g'`) -- what remains is the key id and then the nickname. Take the first field (`awk {print $1}`) as that is the key id we want.